llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1265,6 +1265,128 @@ LogicalResult TopkOp::verify() {
+    }
+    loopBounds[dim.index()].stride = one;
+  }
+  return loopBounds;
+}
+
+SmallVector<StringRef> TopkOp::getLoopIteratorTypes() {
+  SmallVector<StringRef> iteratorTypes(getInputRank(),
+                                       getParallelIteratorTypeName());
+  iteratorTypes[dimension()] = getReductionIteratorTypeName();
+  return iteratorTypes;
+}
+
+LogicalResult TopkOp::generateScalarImplementation(OpBuilder &b, Location loc,
+                                                   ValueRange ivs) {
+  uint64_t kDim = dimension();
+  Value zero = b.create<arith::ConstantIndexOp>(loc, 0);
+  Value one = b.create<arith::ConstantIndexOp>(loc, 1);
+  Value initialValue = b.create<memref::LoadOp>(loc, values(), ivs);
+  Value initialIndex = b.create<memref::LoadOp>(loc, indices(), ivs);
+
+  // Compute K (ub) from the selected dim of the output
+  Value ub;
+  ShapedType outputValuesType = outputValues().getType().cast<ShapedType>();
+  if (outputValuesType.isDynamicDim(dimension())) {
+    ub = b.create<memref::DimOp>(loc, outputValues(), dimension());
+  } else {
+    ub = b.create<arith::ConstantIndexOp>(
+        loc, outputValuesType.getDimSize(dimension()));
+  }
+
+  // Inner K loop functions:
+  //   Load current K val/ind
+  //   Compare N to K using inserted block compare
+  //   Check if N/K are equal, which came first
+  //   Store results in appropriate places
+  //   Yield loop carry vals
+  Value kValue, kIndex;
+  auto scfFor = b.create<scf::ForOp>(
+      loc, zero, ub, one, ValueRange{initialValue, initialIndex},
+      [&](OpBuilder &b, Location loc, Value iv, ValueRange loopCarryValues) {
+        SmallVector<Value> indices(ivs);
+        indices[kDim] = iv;
+        kValue = b.create<memref::LoadOp>(loc, outputValues(), indices);
+        kIndex = b.create<memref::LoadOp>(loc, outputIndices(), indices);
+      });
+
+  SmallVector<Value> indices(ivs);
+  indices[kDim] = scfFor.getInductionVar();
+  auto loopCarryValues = scfFor.getRegionIterArgs();
+
+  // Retrieve region as black box comparision function f(x,y). Plug into op.
+  auto &srcBlock = region().front();
+  BlockAndValueMapping bvmF; // f(x,y)
+  BlockAndValueMapping bvmR; // f(y,x)
+  {
+    // Save previous insertion point. Continue within loop body.
+    OpBuilder::InsertionGuard guard(b);
+    b.setInsertionPointToEnd(&scfFor.getRegion().front());
+    for (auto it : llvm::zip(srcBlock.getArguments(),
+                             ValueRange{loopCarryValues[0], kValue})) {
+      bvmF.map(std::get<0>(it), std::get<1>(it));
+    }
+    for (auto it : llvm::zip(srcBlock.getArguments(),
+                             ValueRange{kValue, loopCarryValues[0]})) {
+      bvmR.map(std::get<0>(it), std::get<1>(it));
+    }
+    for (auto &blockOp : srcBlock.without_terminator()) {
+      b.clone(blockOp, bvmF);
+      b.clone(blockOp, bvmR);
+    }
+    Value forwardCmpRes =
+        bvmF.lookupOrDefault(srcBlock.getTerminator()->getOperand(0));
```

Comment:
I'd rather use `lookup` here. That will assert if it doesnt find the value, which would be the right behavior. I see that above also does `lookupOrDefault` . That should be changed as well (if you change it and nothing breaks, great!, if not you can leave it as is).